staging/src/k8s.io/client-go/tools/watch/retrywatcher.go

```diff
@@ -0,0 +1,179 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package watch
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+// WatcherFunc is a function that is responsible for creating a watcher starting at sinceResourceVersion.
+type WatcherFunc func(sinceResourceVersion string) (watch.Interface, error)
+
+// RetryWatcherError is the type of an event that will be returned if RetryWatcher fails.
+type RetryWatcherError struct {
+	s string
+}
+
+// Error implements error interface.
+func (e RetryWatcherError) Error() string {
+	return e.s
+}
+
+// GetObjectKind implements runtime.Object interface.
+func (e RetryWatcherError) GetObjectKind() schema.ObjectKind { return schema.EmptyObjectKind }
+
+// DeepCopyObject implements runtime.Object interface.
+func (e RetryWatcherError) DeepCopyObject() runtime.Object { return e }
+
+var _ error = RetryWatcherError{}
+var _ runtime.Object = RetryWatcherError{}
+
+// resourceVersionGetter is an interface used to get resource version from events.
+// We can't reuse an interface from meta otherwise it would be a cyclic dependency and we need just this one method
+type resourceVersionGetter interface {
+	GetResourceVersion() string
+}
+
+// RetryWatcher will make sure that in case the underlying watcher is closed (e.g. due to API timeout or etcd timeout)
+// it will get restarted from the last point without the consumer even knowing about it.
+// RetryWatcher does that by inspecting events and keeping track of resourceVersion.
+// Especially useful when using watch.Until where premature termination is causing troubles and flakes.
+// Please note that this is not resilient to ETCD cache not having the resource version anymore - you would need to
+// use Informers for that.
+type RetryWatcher struct {
+	lastResourceVersion string
+	watcherFunc         WatcherFunc
+	resultChan          chan watch.Event
+	stopChan            chan struct{}
+}
+
+// NewRetryWatcher creates a new RetryWatcher.
+// It will make sure that watcher gets restarted in case of recoverable errors.
+// The initialResourceVersion will be given to watchFunc when first called.
+func NewRetryWatcher(initialResourceVersion string, watcherFunc WatcherFunc) *RetryWatcher {
+	rw := &RetryWatcher{
+		lastResourceVersion: initialResourceVersion,
+		watcherFunc:         watcherFunc,
+		stopChan:            make(chan struct{}),
+		resultChan:          make(chan watch.Event, 0),
+	}
+	go rw.receive()
+	return rw
+}
+
+// doReceive returns true when it is done, false otherwise
+func (rw *RetryWatcher) doReceive() bool {
+	watcher, err := rw.watcherFunc(rw.lastResourceVersion)
+	if err != nil {
+		rw.resultChan <- watch.Event{
+			Type:   watch.Error,
+			Object: RetryWatcherError{fmt.Sprintf("RetryWatcher: watcherFunc failed: %v", err)},
```

Comment:
using `apierrors.NewInternalError(fmt.Errorf("RetryWatcher: watcherFunc failed: %v", err)).Status()` instead of RetryWatcherError would be more natural and more likely to be handled/surfaced by watchers (metav1.Status errors are typically what is returned with watch.Error events)